### PR TITLE
[adservice] - add manual tracing instrumentation

### LINKED ISSUE #60: Review and enhance tracing support for ad service service (Java)
The following is a list of requirements that we need to evaluate before declaring v1 for trace telemetry. These requirements are across the entire application; Not all services will meet all requirements. Determine the relevant features for this service. 

- [x] Automatic Instrumentation is being used where appropriate.
- [ ] Library instrumentation is being used if automatic instrumentation is unavailable.
- [ ] #151
  - [x] #152
  - [ ] New spans are being created from existing spans.
- [ ] Automatic and Manual Context Propagation are being demonstrated.
  - [ ] Interprocess context propagation is demonstrated.
- [ ] Telemetry is not being sampled upfront.
- [ ] Telemetry is not being filtered upfront.
- [ ] Baggage is being set and read appropriately (i.e., baggage must be explicitly set as attributes on spans)

Referencing: https://github.com/open-telemetry/opentelemetry-demo-webstore/issues/42

Service: https://github.com/open-telemetry/opentelemetry-demo-webstore/blob/main/src/adservice/README.md

### PR DESCRIPTION
**Pull request by puckpuck**
Fixes #60 

- [x] Services extend automatic instrumentation.
  - [x] New attributes/events attached to existing spans.
  - [x] New spans are being created from existing spans.

## Changes

Uses a mixture of adding spans and attributes using annotations, as well as manual creation of spans.
Attributes are added to existing auto-instrumented spans and newly created manual ones.

### CODE REVIEW

**Review by julianocosta89 — APPROVED** [review]
LGTM!
I've also deployed locally to ensure everything was compiling and working properly!
🚀 it!

**Review by austinlparker — DISMISSED** [review]
Overall lgtm, but I had some questions.

File: src/adservice/src/main/java/hipstershop/AdService.java [hunk 1]
```diff
@@ -106,10 +119,12 @@ public void getAds(AdRequest req, StreamObserver<AdResponse> responseObserver) {
           // Serve random ads.
           allAds = service.getRandomAds();
         }
+        span.setAttribute("app.ads.count", allAds.size());
         AdResponse reply = AdResponse.newBuilder().addAllAds(allAds).build();
         responseObserver.onNext(reply);
         responseObserver.onCompleted();
       } catch (StatusRuntimeException e) {
+        span.setStatus(StatusCode.ERROR);
```
[the remark this answers is not in the document]
Comment (puckpuck) [reply src/adservice/src/main/java/hipstershop/AdService.java:127]: 🤦 When I wrote this code block I was actually telling myself to add a Span event.

Will get it in there before we merge this

File: src/adservice/src/main/java/hipstershop/AdService.java [hunk 2]
```diff
@@ -118,18 +133,36 @@ public void getAds(AdRequest req, StreamObserver<AdResponse> responseObserver) {
...
+    Span span = tracer.spanBuilder("getRandomAds").startSpan();
+
+    // put the span into context, so if any child span is started the parent will be set properly
+    try (Scope ignored = span.makeCurrent()) {
+
+      Collection<Ad> allAds = adsMap.values();
+      for (int i = 0; i < MAX_ADS_TO_SERVE; i++) {
+        ads.add(Iterables.get(allAds, random.nextInt(allAds.size())));
+      }
+      span.setAttribute("app.ads.count", ads.size());
```
[the remark this answers is not in the document]
Comment (puckpuck) [reply src/adservice/src/main/java/hipstershop/AdService.java:160]: There are times you can have more than 1 category, and `getAdsByCategory` will get called multiple times, resulting in different counts.

**Review by wph95 — APPROVED** [review]
just help authoer to fix PR describe todolist format
wrong
```
[x] xxxxx
```

```
- [x]  xxx
```

File: src/adservice/src/main/java/hipstershop/AdService.java [hunk 3]
```diff
@@ -91,8 +100,14 @@ private static class AdServiceImpl extends hipstershop.AdServiceGrpc.AdServiceIm
...
     public void getAds(AdRequest req, StreamObserver<AdResponse> responseObserver) {
       AdService service = AdService.getInstance();
+
+      // get the current span in context
+      Span span = Span.current();
       try {
         List<Ad> allAds = new ArrayList<>();
+
+        span.setAttribute("app.ads.contextKeys", req.getContextKeysList().toString());
+        span.setAttribute("app.ads.contextKeys.count", req.getContextKeysCount());
```
Comment (mic-max) [inline src/adservice/src/main/java/hipstershop/AdService.java:110]: Should we use `app.ads` or `app.adservice` as the namespace

**Review by mic-max — APPROVED** [review]

File: src/adservice/src/main/java/hipstershop/AdService.java (hunk 3 above)
Comment (puckpuck) [reply src/adservice/src/main/java/hipstershop/AdService.java:110]: I would argue against including the `service` suffix in all attribute names.

Comment (puckpuck) [reply src/adservice/src/main/java/hipstershop/AdService.java:110]: Doing instrumentation from other services, we are using attribute names that don't reflect the service they are defined in.  Such as `app.order.id` which is part of the email service. If we are to use the service names in the attribute names, this should be handled in a larger conversation.

**Comment by cartersocha** [comment]
@austinlparker, @mic-max any comments to Pierre's responses ?

File: src/adservice/src/main/java/hipstershop/AdService.java (hunk 3 above)
Comment (cartersocha) [reply src/adservice/src/main/java/hipstershop/AdService.java:110]: What would the intent be behind adding service name? If it's part of a span in a service it's inherently part of that service unless that's not captured in the span info? If it's for querying I feel like adding a single service name attribute is preferable to every attribute having a service prefix

File: src/adservice/src/main/java/hipstershop/AdService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (austinlparker) [reply src/adservice/src/main/java/hipstershop/AdService.java:163]: Ok, sounds good

File: src/adservice/src/main/java/hipstershop/AdService.java (hunk 3 above)
Comment (austinlparker) [reply src/adservice/src/main/java/hipstershop/AdService.java:110]: I don't think we need to namespace the service name in these attribute keys? There's already service.name.

**Comment by cartersocha** [comment]
@puckpuck please update the branch & we can merge

**Pull request merged**